common/filesystem.h

```diff
@@ -0,0 +1,1374 @@
+  auto is_file() const -> bool { return type() == FileType::RegularFile; }
+  auto is_symlink() const -> bool { return type() == FileType::SymbolicLink; }
+
+  // The read, write, and execute permissions for user, group, and others.
+  //
+  // These are represented using the Unix-style bit pattern that facilitates
+  // octal modeling:
+  // - Owner bit mask: 0700
+  // - Group bit mask: 0070
+  // - All bit mask:   0007
+  //
+  // For each, read is an octal value of `1`, write `2`, and execute `4`.
+  //
+  // Windows gracefully degrades to the effective permissions modeled using
+  // these values.
+  auto permissions() const -> ModeType { return stat_buf_.st_mode & 0777; }
+
+  // Non-portable APIs only available on Unix-like systems. See the
+  // documentation of the Unix `stat` structure fields they expose for their
+  // meaning.
+  auto inode() const -> uint64_t { return stat_buf_.st_ino; }
+  auto uid() const -> uid_t { return stat_buf_.st_uid; }
+
+ private:
+  friend DirRef;
+  friend Internal::FileRefBase;
+
+  FileStatus() = default;
+
+  struct stat stat_buf_ = {};
+};
+
+// The base class defining the core `File` API.
+//
+// While not used directly, this is the base class used to implement all of the
+// main `File` types: `ReadFileRef`, `WriteFileRef`, and `ReadWriteFileRef`.
+//
+// Objects using this type have access to an open file handle to a specific file
+// and expose operation on that open file. These operations may fail directly
+// with their `ErrorOr` return, but some errors may be deferred until the
+// underlying owning file is closed.
+//
+// The type provides reference semantics to the underlying file, but is
+// rebindable, movable, and copyable unlike a C++ language reference.
+class Internal::FileRefBase {
+ public:
+  // This object can be default constructed, but will hold an invalid file
+  // handle in that case. This is to support rebinding operations.
+  FileRefBase() = default;
+
+  // These objects are movable and copyable.
+  FileRefBase(FileRefBase&&) = default;
+  FileRefBase(const FileRefBase&) = default;
+  auto operator=(FileRefBase&&) -> FileRefBase& = default;
+  auto operator=(const FileRefBase&) -> FileRefBase& = default;
+
+  // Computers the file status.
+  //
+  // Analogous to the Unix-like `fstat` call.
+  auto Stat() -> ErrorOr<FileStatus, FdError>;
+
+  // Methods to seek the current file position, with various semantics for the
+  // offset.
+  auto Seek(int64_t delta) -> ErrorOr<int64_t, FdError>;
+  auto SeekTo(int64_t pos) -> ErrorOr<int64_t, FdError>;
+  auto SeekFromEnd(int64_t delta) -> ErrorOr<int64_t, FdError>;
+
+  // Reads as much data as is available and fits into the provided buffer.
+  //
+  // On success, this returns a new slice from the start to the end of the
+  // successfully read bytes. These will always be located in the passed-in
+  // buffer, but not all of it may be filled. A partial read does not mean that
```

Comment:
```suggestion
  // buffer, but not all of the buffer may be filled. A partial read does not mean that
```
This reads like not all of the returned slice may be filled, but I don't think it's saying that